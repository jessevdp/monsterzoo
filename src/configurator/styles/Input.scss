@import '~@local/styles/variables/spacing';

.input {
    display: flex;
    flex-direction: column;
    padding: spacing(1) spacing(2) spacing(2);

    > label {
        font-size: .6rem;
        letter-spacing: .1em;
        text-transform: uppercase;
    }

    > input {
        appearance: none;
        background: transparent;
        border: 0;
        box-sizing: border-box;
        margin: 0;
        padding: 0;

        font-size: 1rem;
        margin-top: spacing(1);
        margin-left: spacing(2);

        &:focus,
        &:active {
            outline: 0;
        }
    }
}