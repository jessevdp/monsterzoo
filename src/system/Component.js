import uuid from 'uuid/v1';
import {
    isFunction,
    isString,
    isObject,
    isArray,
} from '@local/utilities';

export default class Component {
    /**
     * Creates an instance of Component.
     * @memberof Component
     */
    constructor() {
        this.id = uuid();
        this.state = {};
        registerEventListeners(this);
        this._observer = registerDOMMutationObserver(this);
    }

    /**
     * Render the view of the component
     *
     * @returns {string} a string representation of the rendered view
     * @memberof Component
     */
    render() {
        verifyHasView(this);
        const view  = this.view();
        verifyView(view);
        return addComponentId(view, this.id);
    }

    /**
     * Re-render the component to the DOM.
     * If the component has been rendered to the DOM before this method can be used to replace
     * the entire component with an updated version.
     * 
     * @returns {void}
     * @memberof Component
     */
    update() {
        const element = this.getHTMLElement();
        if (element) {
            const newElement = toDOMNode(this.render());
            element.replaceWith(newElement);
        }
    }

    afterDOMUpdate() {
        cleanUpEffects(this);
        runEffects(this);
    }

    /**
     * Mutate the state of the component
     *
     * @param {(object|function)} state the subset of the state to be mutated, or a function that returns this subset
     * @returns {void}
     * @memberof Component
     */
    setState(state) {
        let newState;
        if (isObject(state)) newState = state;
        else if (isFunction(state)) newState = state(this.state);
        this.state = { ...this.state, ...newState };
        this.update();
    }

    /**
     * Register an event listener on the component
     *
     * @param {string} event a string representing the event-type to listen for
     * @param {string} [selector=undefined]
     * @param {function} handler
     * @returns {void}
     * @memberof Component
     */
    on(event, ...args) {
        const handler = args.pop();
        const selector = args.pop();
        const query = `[data-component-id="${this.id}"] ${selector ? selector : ''}`;
        const options = { capture: true, passive: true };

        document.addEventListener(event, (e) => {
            if (containsElement(query, e.target)) handler(e);
        }, options);
    }

    /**
     * Get the HTMLElement for the component by querying the document for an element with the
     * id of the component.
     *
     * @returns {(HTMLElement|null)} Returns the first element with the same component id. Or null if not found.
     * @memberof Component
     */
    getHTMLElement() {
        return document.querySelector(`[data-component-id="${this.id}"]`);
    }

    /**
     * Clean up any side effects that the Component has created
     * 
     * @returns {void}
     * @memberof Component
     */
    cleanup() {
        this._observer.disconnect();
    }
}

/**
 * Check to if a query on the document contains a certain element.
 *
 * @param {string} query 
 * @param {HTMLElement} target
 * @returns {boolean}
 */
function containsElement(query, target) {
    const elements = document.querySelectorAll(query);
    let found = false;
    elements.forEach(element => {
        if (element.contains(target)) found = true;
    });
    return found;
}

/**
 * Create a MutationObserver that calls Component.afterDOMUpdate when the
 * mounted DOM representation of the component, or any of it's sub-nodes
 * are updated.
 * 
 * @returns {void}
 * @param {Component} component
 */
function registerDOMMutationObserver(component) {
    const config = { childList: true, subtree: true };
    const observer = new MutationObserver(mutations => {
        const $component = component.getHTMLElement();
        const wasUpdated = mutations
            .map(mutation => mutation.addedNodes)
            .map(nodeList => [...nodeList])
            .reduce((nodes, subset) => nodes.concat(subset), [])
            .map(node => node.contains($component))
            .reduce((matched, matches) => matched || matches, false);
        if (wasUpdated) component.afterDOMUpdate();
    });
    observer.observe(document, config);
    return observer;
}

function runEffects(component) {
    if (!isFunction(component.effects)) return;

    const cleanup = [];

    function useEffect(effect) {
        const result = effect();
        if (isFunction(result)) cleanup.push(result);
    }

    component.effects(useEffect);
    component._toCleanUp = cleanup;
}

function cleanUpEffects(component) {
    if (isArray(component._toCleanUp)) {
        component._toCleanUp.forEach(cleanup => cleanup());
    }
}

function registerEventListeners(component) {
    if (isFunction(component.events)) component.events();
}

function verifyHasView(component) {
    if (!isFunction(component.view)) {
        throw new Error('Component must implement a [view] function.');
    }
}

function verifyView(view) {
    if (!isString(view)) {
        throw new Error('Component [view] function must return a string.');
    }
    const wrapper = wrapHtmlString(view);
    if (wrapper.children.length !== 1) {
        throw new Error('Component [view] function must return exactly 1 root element');
    }
}

function addComponentId(view, id) {
    const wrapper = wrapHtmlString(view);
    const element = wrapper.children[0];
    element.setAttribute('data-component-id', id);
    return wrapper.innerHTML;
}

function wrapHtmlString(htmlString, element = 'div') {
    const wrapper = document.createElement(element);
    wrapper.innerHTML = htmlString;
    return wrapper;
}

function toDOMNode(view) {
    const wrapper = wrapHtmlString(view);
    return wrapper.children[0];
}
