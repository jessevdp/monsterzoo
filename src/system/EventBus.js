import { isArray } from '@local/utilities/';

export default class EventBus {
    constructor() {
        this.events = {};
    }

    /**
     * Subscribe to a certain event with a handler
     *
     * @param {string} event
     * @param {function} handler
     * @returns {void}
     */
    on(event, handler) {
        if (!isArray(this.events[event])) this.events[event] = [];
        this.events[event].push(handler);
    }

    /**
     * Unsubscribe a handler form a certain event
     *
     * @param {string} event
     * @param {function} handler
     * @returns {void}
     */
    remove(event, handler) {
        if (!isArray(this.events[event])) return;
        const index = this.events[event].indexOf(handler);
        if (index > -1) this.events[event].splice(index, 1);
        if (this.events[event].length <= 0) delete this.events[event];
    }

    /**
     * Emit an event, passing optional data to every listener/handler.
     *
     * @param {string} event
     * @param {*} [args]
     * @returns {void}
     */
    emit(event, ...args) {
        if (!isArray(this.events[event])) return;
        this.events[event].forEach(handler => handler(...args));
    }
}