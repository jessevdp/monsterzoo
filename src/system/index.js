/* eslint-env node */
const system = require('./system');
module.exports = system.default || system;