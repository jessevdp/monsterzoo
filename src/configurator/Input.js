import { Component } from '@local/system';

const defaultAttributes = {
    'type': 'text'
};

export default class Input extends Component {
    /**
     * Creates an instance of Input.
     * @param {string} name
     * @param {object} [attributes={}] A set of attributes for the input box, represented as key-value pairs
     * @memberof Input
     */
    constructor(name, attributes = {}) {
        super();
        attributes = {...defaultAttributes, ...attributes};
        this.setState({
            name,
            attributes
        });
    }

    /**
     * Convert the attributes object (stored in state) into it's HTML string
     * equivalent. (E.g. a sequence of: 'key="value"')
     *
     * @returns {string}
     * @memberof Input
     */
    htmlAttributes() {
        return Object.entries(this.state.attributes)
            .map((entry) => `${entry[0]}="${entry[1]}"`)
            .reduce((all, attribute) => all + (attribute + ' '), '')
            .trim();
    }
}