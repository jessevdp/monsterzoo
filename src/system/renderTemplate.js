import mustache from 'mustache';
import { isObject, isArray } from '@local/utilities';
import Component from './Component';

/**
 * Render a template string with a given data object.
 *
 * @export
 * @param {string} template the template string
 * @param {object} [data={}] the view-data object
 * @returns {string} the rendered template populated with the data
 */
export default function renderTemplate (template, data = {}) {
    if (typeof template !== 'string') throw new Error('Invalid [template] parameter, expected a string.');
    if (!isObject(data)) throw new Error('Invalid [data] parameter, expected an object.');
    return mustache.render(template, renderComponentsInObject(data));
}

/**
 * Replaces instances of Component with their rendered version
 *
 * @param {object} object
 * @returns {object}
 */
function renderComponentsInObject(object) {
    const newObject = {};
    for (let key in object) {
        const current = object[key];
        if (current instanceof Component) newObject[key] = current.render();
        else if (isObject(current)) newObject[key] = renderComponentsInObject(current);
        else if (isArray(current)) newObject[key] = renderComponentsInArray(current);
        else newObject[key] = object[key];
    }
    return newObject;
}

/**
 * Replaces instances of Component with their rendered version
 *
 * @param {array} array
 * @returns {array}
 */
function renderComponentsInArray(array) {
    return array.map(item => {
        if (item instanceof Component) return item.render();
        else return item;
    });
}
