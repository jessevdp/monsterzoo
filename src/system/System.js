import Component from './Component';
import renderTemplate  from './renderTemplate';

const System = {
    Component,

    renderTemplate,
};

export default System;
