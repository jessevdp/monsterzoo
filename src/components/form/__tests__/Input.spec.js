import Input from '../Input';
import { Component } from '@local/system';

beforeEach(() => {
    document.body.innerHTML = '';
});

it('renders correctly', () => {
    const input = new Input('name', 'label');
    input.state.attributes = { foo: 'bar', bar: 'foo' };
    expect(input.view()).toMatchSnapshot();
    input.cleanup();
})

describe('constructor', () => {
    it('sets the [name] on state', () => {
        const name = 'foobar';
        const input = new Input(name, 'label');
        expect(input.state.name).toBe(name);
        input.cleanup();
    })
    it('sets the [label] on state', () => {
        const label = 'foobar';
        const input = new Input('name', label);
        expect(input.state.label).toBe(label);
        input.cleanup();
    })
    it('uses defaults for the [attributes]', () => {
        const input = new Input('foobar');
        expect(input.state.attributes).toMatchSnapshot();
        input.cleanup();
    })
    it('merges the passed attributes with the defaults', () => {
        const input = new Input('name', 'label', { foo: 'bar' });
        expect(input.state.attributes).toHaveProperty('foo', 'bar');
        input.cleanup();
    })
    it('overwrites attribute defaults with passed attributes', () => {
        const input = new Input('name', 'label', { type: 'number' });
        expect(input.state.attributes.type).toBe('number');
        input.cleanup();
    })
    it('accepts an optional [value] attribute on [attributes] to initialze state.value', () => {
        const value = 'value';
        const attributes = { foo: 'bar', value };
        const input = new Input('name', 'label', attributes);
        expect(input.state.value).toBe(value);
        expect(input.state.attributes.foo).toBe(attributes.foo);
        expect(input.state.attributes.value).toBe(undefined);
        input.cleanup();
    })
})

describe('htmlAttributes', () => {
    it('can convert [attributes] state object into an HTML string', () => {
        const attributes = { foo: 'bar', number: 10 }
        const expected = 'foo="bar" number="10"';
        const input = new Input('name');
        input.state.attributes = attributes;
        expect(input.htmlAttributes()).toBe(expected);
        input.cleanup();
    })
})

describe('setAttributes', () => {
    it('accepts an object to update the attributes', () => {
        const input = new Input('name');
        const attributes = { foo: 'bar' };
        input.setAttributes(attributes);
        expect(input.state.attributes.foo).toBe(attributes.foo);
        input.cleanup();
    })
    it('accepts a function to update the attributes', () => {
        const input = new Input('name');
        const attributes = { foo: 'bar' };
        input.setAttributes(() => attributes);
        expect(input.state.attributes.foo).toBe(attributes.foo);
        input.cleanup();
    })
    it('passes the old attributes to the passed function', () => {
        // Arrange
        const oldAttributes = { foo: 'foo' };
        const newAttributes = { foo: 'bar' };
        const fn = jest.fn(() => newAttributes);
        const input = new Input('name');
        input.state.attributes = oldAttributes;

        // Act
        input.setAttributes(fn);

        // Assert
        expect(fn).toHaveBeenCalledWith(oldAttributes);

        // Cleanup
        input.cleanup();
    })
    it('overwrites old values', () => {
        const oldAttributes = { foo: 'foo' };
        const newAttributes = { foo: 'bar' };
        const input = new Input('name', oldAttributes);
        input.setAttributes(newAttributes);
        expect(input.state.attributes.foo).toBe(newAttributes.foo);
        input.cleanup();
    })
    it('leaves other values intact', () => {
        const input = new Input('name');
        const fooAttribute = { foo: 'foo' };
        const barAttribute = { bar: 'bar' };
        input.setAttributes(fooAttribute);
        input.setAttributes(barAttribute);
        expect(input.state.attributes.foo).toBe(fooAttribute.foo);
        expect(input.state.attributes.bar).toBe(barAttribute.bar);
        input.cleanup();
    })
})

describe('[label] setter', () => {
    it('updates the [label] variable on state', () => {
        const newLabel = 'new label';
        const input = new Input('name', 'label');
        input.label = newLabel;
        expect(input.state.label).toBe(newLabel);
        input.cleanup();
    })
    it('uses the setState method', () => {
        const newLabel = 'new label';
        const input = new Input('name', 'label');
        const spy = jest.spyOn(Component.prototype, 'setState');
        input.label = newLabel;
        expect(spy).toHaveBeenCalledWith({ label: newLabel });
        input.cleanup();
    })
})

describe('[value] setter', () => {
    it('updates the [value] variable on state', () => {
        const newValue = 'new value';
        const input = new Input('name', 'label', 'value');
        input.value = newValue;
        expect(input.state.value).toBe(newValue);
        input.cleanup();
    })
    it('updates the value of the input box', () => {
        const newValue = 'new value';
        const input = new Input('name', 'label');
        document.body.innerHTML = input.render();
        input.value = newValue;
        const $input = document.querySelector('input');
        expect($input.value).toBe(newValue);
        input.cleanup();
    })
    it('uses the setState method', () => {
        const newValue = 'new value';
        const input = new Input('name', 'label');
        const spy = jest.spyOn(Component.prototype, 'setState');
        input.value = newValue;
        expect(spy).toHaveBeenCalledWith({ value: newValue });
        input.cleanup();
    })
})
describe('[value] getter', () => {
    it('returns the value of state.value', () => {
        const value = 'value';
        const input = new Input('name', 'label');
        input.state.value = value;
        expect(input.value).toBe(value);
        input.cleanup();
    })
})

describe('events', () => {
    it('updates [value] state when the value of the input changes', () => {
        // Arrange
        const newValue = 'new value';
        const input = new Input('name', 'label');
        document.body.innerHTML = input.render();
        const $input = document.querySelector('input');

        // Act
        $input.value = newValue;
        $input.dispatchEvent(new Event('change'));

        // Assert
        expect(input.value).toBe(newValue);

        // Cleanup
        input.cleanup();
    })
})