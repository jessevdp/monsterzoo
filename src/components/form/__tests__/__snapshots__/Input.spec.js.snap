// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructor uses defaults for the [attributes] 1`] = `
Object {
  "type": "text",
}
`;

exports[`renders correctly 1`] = `"<div class=\\"input bg-300\\"><label class=\\"text-secondary\\">label</label><input name=\\"name\\" value=\\"\\" foo=\\"bar\\" bar=\\"foo\\" /></div>"`;
